Simplify readme (#5217)
koreader/koreader: doc/Building.md

# Setting a build environment for KOReader

> These instructions are intended to build the emulator in Linux and MacOS. Windows users are suggested to develop in a Linux VM.

Author: I restored the link in Linux VM and added a wikipedia link for WSL.

If you only want to work with Lua frontend stuff, you can grab the AppImage and
run it with `--appimage-extract`.

> ## Prerequisites

Author: Yeah, I just copied and pasted. Any additional info would be great!

# Setting up a build environment for KOReader

> These instructions are intended to build the emulator in Linux and MacOS. Windows users are suggested to develop in a [Linux VM](https://www.howtogeek.com/howto/11287/how-to-run-ubuntu-in-windows-7-with-vmware-player/) or using the [Windows Subsystem for Linux](https://en.wikipedia.org/wiki/Windows_Subsystem_for_Linux).

Reviewer: WSL was more thinking aloud. I'm not sure we should recommend using WSL unless someone's actually verified it works for the purpose? The AppImage runs in WSL so logic dictates it should, but... :-)

Not that I've verified VMWare player personally or anything, but it's Ubuntu ergo it will work as expected.

Since I happen to be typing in Windows right now I suppose I'll give [this](https://docs.microsoft.com/en-us/windows/wsl/wsl2-install) a go, but it wants to restart after executing the first command and I don't want to restart. :-P

If you only want to work with Lua frontend stuff, you can grab the AppImage and
run it with `--appimage-extract`.

You can skip most of the following instructions if desired, and use our premade Docker image instead. In that case the only requirements are Git and Docker. See [the virtual development environment README](https://github.com/koreader/virdevenv) for more information.

## Prerequisites

To get and compile the source you must have `patch`, `wget`, `unzip`, `git`,

> `cmake` and `luarocks` installed, as well as a version of `autoconf`

Reviewer: luarocks seems to have become a minor issue in the last year (regardless whether we're talking Mac OS or bleeding user distros like Arch & Gentoo). For Docker we roll out our own, see https://github.com/koreader/virdevenv/blob/9b41d9f36d9b20fdf9b58d04104cc9133fc15ce5/docker/ubuntu/baseimage/install_luarocks.sh#L7-L13

We should probably at least add a little note to that script for the moment, although perhaps ideally e'd move away from system luarocks since its config files/options and such are annoyingly variable across versions.

Author: We can address this issue here too: https://github.com/koreader/koreader/issues/2798#issuecomment-299973036

greater than 2.64. You also need `nasm` and of course a compiler like `gcc`
or `clang`. If you want to cross-compile for other architectures, you need a proper

> cross-compile toolchain. Your GCC should be at least version 4.8.

Reviewer: ```suggestion
cross-compile toolchain. Your GCC should be at least version 4.9.
```

This must be duplicated? :-)

Author: Yes :p

### Debian/Ubuntu and derivates

> Install the prerequisites using dpkg:

Reviewer: dpkg? Perhaps apt uses `dpkg -i` behind the scenes but for an aspiring hacker that's just adding unnecessary confusion. Here it says dpkg, then it says apt-get? Where does dpkg come in?
```suggestion
Install the prerequisites using APT:
```

```
sudo apt-get install build-essential git patch wget unzip \
gettext autoconf automake cmake libtool nasm luarocks libsdl2-dev \
libssl-dev libffi-dev libsdl2-dev libc6-dev-i386 xutils-dev linux-libc-dev:i386 zlib1g:i386
```

### Fedora/Red Hat

> Install `libstdc++-static`, `SDL` and `SDL-devel` packages using `yum`

Reviewer: Yum has been replaced by DNF. I don't know the specifics because I don't really use Fedora. I think the commands are essentially the same, just with `dnf` instead of `yum`. Not that we're giving any actual commands here, lol.

```suggestion
Install `libstdc++-static`, `SDL` and `SDL-devel` packages using DNF:

```
sudo dnf install libstdc++-static SDL SDL-devel
```

Not 100 % sure that'll actually work on post-2014/2015 Fedora. Package names might've changed. I'll probably check in a VM (or possibly on bare metal) later.

### MacOS

Install the prerequisites using [Homebrew](https://brew.sh/):

```
brew install nasm binutils libtool autoconf automake cmake makedepend \
sdl2 lua@5.1 luarocks gettext pkg-config wget md5sha1sum
echo 'export PATH="/usr/local/opt/gettext/bin:$PATH"' >> "$HOME"/.bash_profile
```

If you run into a gettext error while building glib, try `brew link --force gettext` to override the built-in Mac OS BSD gettext with GNU GetText.

*Note:* in Mojave (10.14) you need to set a minimum deployment version higher than 10.04. Otherwise you'll get the error `ld: library not found for -lgcc_s.10.4`.
```
export MACOSX_DEPLOYMENT_TARGET=10.09
```

## Getting the source

```
git clone https://github.com/koreader/koreader.git
cd koreader && ./kodev fetch-thirdparty
```

Building the emulator

## Building and running the emulator

> To build an emulator on your current Linux or MacOS machine:

Reviewer: Current? As opposed to what? :-P
```suggestion
To build an emulator on your Linux or MacOS machine:
```

```
./kodev build
```

To run KOReader on your development machine:

```
./kodev run
```

You can specify the size and DPI of the emulator's screen using
`-w=X` (width), `-h=X` (height), and `-d=X` (DPI).

 There is also a convenience
`-s` (simulate) flag with some presets like `kobo-aura-one`, `kindle3`, and
`hidpi`. The latter is a fictional device with `--screen_width=1500`,
`--screen_height=2000` and `--screen_dpi=600` to help ensure DPI scaling works correctly.

Sample usage:

```
./kodev run -s=kobo-aura-one
```

To use your own koreader-base repo instead of the default one change the `KOR_BASE`
environment variable:

```
make KOR_BASE=../koreader-base
```

> This will be handy if you are developing `koreader-base` and you want to test your

Reviewer: Not sure I disagree with this; just using branches in your base submodule works perfectly fine. But it's potentially slightly more convenient to quickly switch back & forth between different experiments, particularly if they involve Clang vs GCC or something drastic like that.

Author: I never used that, but as you can imagine I copied blindly almost all the information and only tried to digest the doc for final users.

modifications with the KOReader frontend. NOTE: this only supports relative path for now.

## Building for other platforms

Once you have the emulator ready to rock you can [build for other platforms too](Building_targets.md).

## Testing

> You may need to checkout the [circleci config file][circleci-conf] to setup up

Reviewer: ```suggestion
You may need to check out the [circleci config file][circleci-conf] to setup up
```

a proper testing environment. 

Briefly, you need to install `luarocks` and then install `busted` with `luarocks`. The "eng" language data file for tesseract-ocr is also need to test OCR functionality. Finally, make sure that `luajit` in your system is at least of version 2.0.2.

To automatically set up a number of primarily luarocks-related environment variables:

```
./kodev activate
```

To run unit tests:

```
./kodev test base
./kodev test front
```

To run a specific unit test (for test development):

```
./kodev test front readerbookmark_spec.lua
```

> NOTE: Extra dependencies for tests: `busted` and `ansicolors` from luarocks.

Reviewer: Seems a bit double with above on line 109.

To run Lua static analysis:

```
make static-check
```

> NOTE: Extra dependencies for tests: `luacheck` from luarocks

Reviewer: ```suggestion
NOTE: Extra dependencies for tests: `luacheck` from luarocks.
```

## Translations

Please refer to [l10n's README][l10n-readme] to grab the latest translations
from [the KOReader project on Transifex][koreader-transifex] with this command:

```
make po
```

If your language is not listed on the Transifex project, please don't hesitate
to send a language request [here][koreader-transifex].

### Variables in translation

> Some strings contain variables that should remain unaltered in translation.

Reviewer: We should add a little extra warning. Not sure how much good it'll do, but it's happened. ;-)
```suggestion
Some strings contain variables that should remain unaltered in translation. These take the form of a `%` followed by a number from `1-99`, although you'll seldom see more than about 5 in practice. Please don't put any spaces between the `%` and its number. `%1` should always remain `%1`.
```

For example:

```lua
The title of the book is %1 and its author is %2.
```

This might be displayed as:

```lua
The title of the book is The Republic and its author is Plato.
```

To aid localization the variables may be freely positioned:

```lua
De auteur van het boek is %2 en de titel is %1.
```

That would result in:

```lua
De auteur van het boek is Plato en de titel is The Republic.
```

## Use ccache

Ccache can speed up recompilation by caching previous compilations and detecting
when the same compilation is being repeated. In other words, it will decrease
build time when the sources have been built before. Ccache support has been added to
KOReader's build system. To install ccache:

* in Ubuntu use:`sudo apt-get install ccache`

> * in Fedora use:`sudo yum install ccache`

Reviewer: ```suggestion
* in Fedora use:`sudo dnf install ccache`
```
